checked/types.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package checked
+
+// Finalizer finalizes a checked resource.
+type Finalizer interface {
+	Finalize()
+}
+
+// FinalizerFn is a function literal that is a finalizer.
+type FinalizerFn func()
+
+// Finalize will call the function literal as a finalizer.
+func (fn FinalizerFn) Finalize() {
+	fn()
+}
+
+// Ref is an entity that checks ref counts.
+type Ref interface {
+	// IncRef increments the ref count to this entity.
+	IncRef()
```

Comment:
I see, this feels intrusive though because developers would need to remember to follow this workflow mentally when switching from the unchecked pools to checked pools, or else their code would panic in unexpected places. Ideally we could make the APIs more friendly to developers. Do you think that'll be achievable, or are we going to chalk this up to the overhead of using checked pools?